src/config/config.rs

```diff
@@ -17,37 +17,53 @@ use std::{
     str::FromStr,
 };
 
-fn canonicalize(path: &Path) -> String {
-    let mut parts: Vec<String> = Vec::new();
+/// Performs canonicalization by removing `foo/../` and `./` components
+/// from `path`, without hitting the file system. It does not turn relative
+/// paths into absolute paths.
+fn normalize_path(path: impl AsRef<Path>) -> PathBuf {
+    let mut parts: Vec<Component<'_>> = vec![];
 
-    for component in path.components() {
+    for component in path.as_ref().components() {
         match component {
-            Component::RootDir | Component::CurDir => {}
+            Component::CurDir => {}
             Component::ParentDir => {
-                parts.pop();
+                parts
+                    .pop()
+                    .expect("Cannot navigate outside of base directory!");
```

Comment:
Instead of panicking we can also `parts.push(c)`. It should not be forbidden to have `../` at the beginning of the string, but then we should make sure that the thing we're `pop`'ing in subsequent iterations is _not_ a `ParentDir` :)

Reply:
Please remove the `expect` part. This is mostly about display after all. One thing I just thought: `/../whatever` should be transformed into `/whatever`, no `whatever`.

Reply:
> Please remove the `expect` part. This is mostly about display after all.

I strongly dislike such inconsistencies, and making such assumptions by imposing arbitrary undocumented limitations on the future. There's nothing preventing this from being used in more serious usecases later on.

> One thing I just thought: `/../whatever` should be transformed into `/whatever`, no `whatever`.

Correct, apparently `/..` is valid... Though for our intents a relative path outside of the path we know about should not! Otherwise we might as well canonicalize the root path and gir directory, and remove the former as prefix from the latter, that solves the problem in the same way.
Anyway, this is now fixed.

Reply:
> I strongly dislike such inconsistencies, and making such assumptions by imposing arbitrary undocumented limitations on the future. There's nothing preventing this from being used in more serious usecases later on.

I don't see any case where it would extended. The current case can be seen in https://github.com/gtk-rs/gtk-rs/pull/366. If we keep this expect, all our `Gir.toml` files become invalid because they all use a path starting with "..". Again: this is about display **only**.

Reply:
> If we keep this expect, all our `Gir.toml` files become invalid because they all use a path starting with "..".

Negative. The path entering this function is the subdirectory concatenated with the gir directory:

```console
foo:gtk-rs/ $ ./generator.py 2>&1 | rg as_ref
[src/config/config.rs:26] path.as_ref() = "gdk/../gir-files"
[src/config/config.rs:26] path.as_ref() = "gdk/sys/../../gir-files"
[src/config/config.rs:26] path.as_ref() = "pango/../gir-files"
[src/config/config.rs:26] path.as_ref() = "pango/sys/../../gir-files"
[src/config/config.rs:26] path.as_ref() = "gdkx11/../gir-files"
[src/config/config.rs:26] path.as_ref() = "gdkx11/sys/../../gir-files"
[src/config/config.rs:26] path.as_ref() = "graphene/../gir-files"
[src/config/config.rs:26] path.as_ref() = "graphene/sys/../../gir-files"
[src/config/config.rs:26] path.as_ref() = "gtk/../gir-files"
[src/config/config.rs:26] path.as_ref() = "gtk/sys/../../gir-files"
[src/config/config.rs:26] path.as_ref() = "gio/../gir-files"
[src/config/config.rs:26] path.as_ref() = "gio/sys/../../gir-files"
[src/config/config.rs:26] path.as_ref() = "gdk-pixbuf/../gir-files"
[src/config/config.rs:26] path.as_ref() = "gdk-pixbuf/sys/../../gir-files"
[src/config/config.rs:26] path.as_ref() = "glib/../gir-files"
[src/config/config.rs:26] path.as_ref() = "glib/../gir-files"
[src/config/config.rs:26] path.as_ref() = "glib/sys/../../gir-files"
[src/config/config.rs:26] path.as_ref() = "glib/gobject-sys/../../gir-files"
[src/config/config.rs:26] path.as_ref() = "pangocairo/../gir-files"
[src/config/config.rs:26] path.as_ref() = "pangocairo/sys/../../gir-files"
[src/config/config.rs:26] path.as_ref() = "atk/../gir-files"
[src/config/config.rs:26] path.as_ref() = "atk/sys/../../gir-files"
```

The only breakage is observed in the CI, where `gir-*.toml` files are specified directly in the `tests/sys` (so no subdirectory prepended), and the gir files are cloned at `../gir-files`.

> Again: this is about display **only**.

For now.

Reply:
So if you run your change on the current gtk-rs, is it working?

Reply:
> So if you run your change on the current gtk-rs, is it working?

Correct, it was also working with the `expect` (because none of the paths above try to walk outside of their directory).

Reply:
Well then, no problem! We can keep the `expect`. :)

Reply:
Well, it's already fixed to not need the expect...

And it breaks the tests in `tests/src/test.sh`, as mentioned below.